File: sdk-python/src/agentincident/recorder.py
from __future__ import annotations

import functools
import json
from datetime import datetime, timezone
from typing import Any, Callable

from .hash import compute_hash
from .schema import validate as schema_validate
from .types import Constraint, Incident, Meta, TraceEntry


class Recorder:
    """Record agent tool calls and build AgentIncident records.

    Supports three modes:
    - Context manager: ``with Recorder() as rec: ...``
    - Decorator: ``@rec.trace``
    - Manual: ``rec.record(tool, input, output)``
    """

    def __init__(
        self,
        agent: str | None = None,
        framework: str | None = None,
        environment: str | None = None,
        trigger: str | None = None,
    ) -> None:
        self._entries: list[TraceEntry] = []
        self._seq = 0
        self._agent = agent
        self._framework = framework
        self._environment = environment
        self._trigger = trigger
        self._started_at: datetime | None = None

    def __enter__(self) -> Recorder:
        self._started_at = datetime.now(timezone.utc)
        return self

    def __exit__(self, *exc: Any) -> None:
        pass

    @property
    def entries(self) -> list[TraceEntry]:
        return list(self._entries)

    def record(
        self,
        tool: str,
        input: dict[str, Any],
        output: dict[str, Any],
        irreversible: bool = False,
        agent_reasoning: str | None = None,
        duration_ms: int | None = None,
    ) -> TraceEntry:
        """Record a tool call. Auto-computes seq, ts, and hash."""
        self._seq += 1
        entry = TraceEntry(
            seq=self._seq,
            tool=tool,
            input=input,
            output=output,
            ts=datetime.now(timezone.utc),
            hash=compute_hash(input, output),
            duration_ms=duration_ms,
            irreversible=irreversible,
            agent_reasoning=agent_reasoning,
        )
        self._entries.append(entry)
        return entry

    def trace(
        self,
        func: Callable[..., Any] | None = None,
        *,
        tool: str | None = None,
        irreversible: bool = False,
    ) -> Any:
        """Decorator that wraps a function call as a trace entry.

        Usage:
            @rec.trace
            def my_tool(x, y): ...

            @rec.trace(tool="custom_name", irreversible=True)
            def my_tool(x, y): ...
        """
        def decorator(fn: Callable[..., Any]) -> Callable[..., Any]:
            @functools.wraps(fn)
            def wrapper(*args: Any, **kwargs: Any) -> Any:
                tool_name = tool or fn.__name__
                input_data = kwargs if kwargs else {"args": list(args)} if args else {}
                start = datetime.now(timezone.utc)
                try:
                    result = fn(*args, **kwargs)
                except Exception as exc:
                    elapsed = int(
                        (datetime.now(timezone.utc) - start).total_seconds() * 1000
                    )
                    self.record(
                        tool=tool_name,
                        input=input_data,
                        output={"error": str(exc)},
                        irreversible=irreversible,
                        duration_ms=elapsed,
                    )
                    raise
                elapsed = int(
                    (datetime.now(timezone.utc) - start).total_seconds() * 1000
                )
                output_data = (
                    result if isinstance(result, dict) else {"result": result}
                )
                self.record(
                    tool=tool_name,
                    input=input_data,
                    output=output_data,
                    irreversible=irreversible,
                    duration_ms=elapsed,
                )
                return result
            return wrapper

        if func is not None:
            return decorator(func)
        return decorator

    def to_incident(
        self,
        fault_class: str,
        impact_score: int,
        loss_amount_usd: float | None,
        constraints: list[Constraint],
        **kwargs: Any,
    ) -> Incident:
        """Build an Incident from the recorded trace and metadata."""
        now = datetime.now(timezone.utc)
        meta = Meta(
            schema="agentincident/v0.1",
            created_at=self._started_at or now,
            updated_at=now,
            agent=self._agent,
            framework=self._framework,
            environment=self._environment,
            trigger=self._trigger,
        )
        return Incident(
            trace=list(self._entries),
            constraints=constraints,
            fault_class=fault_class,
            impact_score=impact_score,
            loss_amount_usd=loss_amount_usd,
            meta=meta,
            **kwargs,
        )

    def to_dict(self) -> dict[str, Any]:
        """Export recorded entries as a dict (trace only, no incident wrapper)."""
        from .export import _trace_entry_to_dict
        return {"trace": [_trace_entry_to_dict(e) for e in self._entries]}

    def to_json(self, indent: int = 2) -> str:
        """Export recorded entries as JSON."""
        return json.dumps(self.to_dict(), indent=indent, default=str)

    def validate(self, level: str = "core") -> list[str]:
        """Validate the current state. Requires a full incident for meaningful validation."""
        return schema_validate(self.to_dict(), level=level)

File: sdk-typescript/src/recorder.ts
import { computeHash } from "./hash.js";
import type {
  Constraint,
  Classification,
  Environment,
  FaultClass,
  ImpactScore,
  Incident,
  Liability,
  LossEstimate,
  Meta,
  Response,
  TraceEntry,
} from "./types.js";

export interface RecorderOptions {
  agent?: string;
  framework?: string;
  environment?: Environment;
  trigger?: string;
}

export class Recorder {
  private _entries: TraceEntry[] = [];
  private _seq = 0;
  private _incident: Incident | null = null;
  private _options: RecorderOptions;

  constructor(options?: RecorderOptions) {
    this._options = options ?? {};
  }

  record(
    tool: string,
    input: Record<string, unknown>,
    output: Record<string, unknown>,
    options?: {
      irreversible?: boolean;
      agent_reasoning?: string;
      duration_ms?: number;
    }
  ): TraceEntry {
    this._seq += 1;
    const entry: TraceEntry = {
      seq: this._seq,
      tool,
      input,
      output,
      ts: new Date().toISOString(),
      hash: computeHash(input, output),
    };
    if (options?.duration_ms !== undefined) {
      entry.duration_ms = options.duration_ms;
    }
    if (options?.irreversible !== undefined) {
      entry.irreversible = options.irreversible;
    }
    if (options?.agent_reasoning !== undefined) {
      entry.agent_reasoning = options.agent_reasoning;
    }
    this._entries.push(entry);
    return entry;
  }

  wrap<TArgs extends unknown[], TResult>(
    tool: string,
    fn: (...args: TArgs) => TResult | Promise<TResult>,
    options?: { irreversible?: boolean }
  ): (...args: TArgs) => Promise<TResult> {
    return async (...args: TArgs): Promise<TResult> => {
      const inputObj: Record<string, unknown> = { args };
      const start = Date.now();
      const result = await fn(...args);
      const duration_ms = Date.now() - start;
      const outputObj: Record<string, unknown> = { result };
      this.record(tool, inputObj, outputObj, {
        duration_ms,
        irreversible: options?.irreversible,
      });
      return result;
    };
  }

  toIncident(params: {
    fault_class: FaultClass;
    impact_score: ImpactScore;
    loss_amount_usd: number | null;
    constraints: Constraint[];
    meta?: Meta;
    classification?: Classification;
    liability?: Liability;
    loss_estimate?: LossEstimate;
    response?: Response;
  }): Incident {
    const meta: Meta = {
      schema: "agentincident/v0.1",
      created_at: new Date().toISOString(),
      ...this._options,
      ...params.meta,
    };

    const incident: Incident = {
      trace: [...this._entries],
      constraints: params.constraints,
      fault_class: params.fault_class,
      impact_score: params.impact_score,
      loss_amount_usd: params.loss_amount_usd,
      meta,
    };

    if (params.classification) incident.classification = params.classification;
    if (params.liability) incident.liability = params.liability;
    if (params.loss_estimate) incident.loss_estimate = params.loss_estimate;
    if (params.response) incident.response = params.response;

    this._incident = incident;
    return incident;
  }

  toJSON(indent?: number): string {
    if (!this._incident) {
      throw new Error(
        "No incident created yet. Call toIncident() first."
      );
    }
    return JSON.stringify(this._incident, null, indent);
  }

  validate(level: "core" | "standard" | "full" = "core"): string[] {
    const errors: string[] = [];

    if (this._entries.length === 0) {
      errors.push("trace must contain at least one entry");
    }

    if (!this._incident) {
      errors.push("No incident created yet. Call toIncident() first.");
      return errors;
    }

    const inc = this._incident;

    if (!inc.fault_class) errors.push("fault_class is required");
    if (!inc.impact_score) errors.push("impact_score is required");
    if (inc.loss_amount_usd === undefined) {
      errors.push("loss_amount_usd is required (can be null)");
    }
    if (!inc.constraints || inc.constraints.length === 0) {
      errors.push("constraints must contain at least one entry");
    }

    if (level === "standard" || level === "full") {
      if (!inc.meta) errors.push("meta is required for standard level");
      if (!inc.classification)
        errors.push("classification is required for standard level");
      if (!inc.loss_estimate)
        errors.push("loss_estimate is required for standard level");
    }

    if (level === "full") {
      if (!inc.liability)
        errors.push("liability is required for full level");
      if (!inc.response)
        errors.push("response is required for full level");
      if (inc.response) {
        if (!inc.response.actions)
          errors.push("response.actions is required for full level");
        if (!inc.response.constraint_update)
          errors.push(
            "response.constraint_update is required for full level"
          );
        if (!inc.response.verification)
          errors.push("response.verification is required for full level");
      }
    }

    return errors;
  }

  get entries(): ReadonlyArray<TraceEntry> {
    return this._entries;
  }
}
